### PR TITLE
Move EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing to class.cpp

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Move `EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing` to class.cpp and out of fieldmarshaler.cpp.

This change co-locates `EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing` with the rest of the implementation of `EEClassLayoutInfo`. Additionally, this PR separates out the field blittability check into a separate function instead of burying it in `CollectLayoutFieldMetadataThrowing`.

Finally, it adds a small optimization in the field marshaler implementations where if a field's native size is statically known, the code returns that value instead of calling into one of the FieldMarshaler "virtual" calls.

Originally part of #21415, but extracted out to make that PR smaller.

### CODE REVIEW

File: src/vm/fieldmarshaler.h [hunk 1]
```diff
@@ -89,6 +88,44 @@ enum NStructFieldType
 #define DEFAULT_PACKING_SIZE 32
 
 
+typedef enum
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.h]: If we are going to define the `|=` operator, lets make this a proper `enum class`.

File: src/vm/class.cpp [hunk 2]
```diff
@@ -3195,3 +3195,706 @@ void EEClass::SetPackableField(EEClassFieldId eField, DWORD dwValue)
...
+// Called from the clsloader to load up and summarize the field metadata
+// for layout classes.
+//
+// Warning: This function can load other classes (esp. for nested structs.)
+//=======================================================================
+#ifdef _PREFAST_
+#pragma warning(push)
+#pragma warning(disable:21000) // Suppress PREFast warning about overly large function
+#endif
+VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
```
Comment (AaronRobinsonMSFT) [inline src/vm/class.cpp]: The perf overhead of just this call is probably amazing - so many arguments.

File: src/vm/class.cpp [hunk 3]
```diff
@@ -3195,3 +3195,706 @@ void EEClass::SetPackableField(EEClassFieldId eField, DWORD dwValue)
...
+
+    HRESULT             hr;
+    MD_CLASS_LAYOUT     classlayout;
+    mdFieldDef          fd;
+    ULONG               ulOffset;
+    ULONG cFields = 0;
+
+    // Running tote - if anything in this type disqualifies it from being ManagedSequential, somebody will set this to TRUE by the the time
+    // function exits.
+    BOOL                fDisqualifyFromManagedSequential = FALSE;
```
Comment (AaronRobinsonMSFT) [inline src/vm/class.cpp]: Move this down to the if where it can be initially set instead of just commented about.

File: src/vm/class.cpp (hunk 2 above)
Comment (jkoritzinsky) [reply src/vm/class.cpp]: Yeah and it's hit on type load for every type with non-auto layout. And `ParseNativeType` (with 11 arguments in release, 14 in debug) gets hit once for every instance field of every loaded type with non-auto layout.

File: src/vm/class.cpp [hunk 4]
```diff
@@ -3195,3 +3195,703 @@ void EEClass::SetPackableField(EEClassFieldId eField, DWORD dwValue)
...
+    // function exits.
+    BOOL fDisqualifyFromManagedSequential = FALSE; 
+
+    // Check if this type might be ManagedSequential. Only valuetypes marked Sequential can be
+    // ManagedSequential. Other issues checked below might also disqualify the type.
+    if ( (!fExplicitOffsets) &&    // Is it marked sequential?
+         (pParentMT && (pParentMT->IsValueTypeClass() || pParentMT->IsManagedSequential()))  // Is it a valuetype or derived from a qualifying valuetype?
+       )
+    {
+        // Type qualifies so far... need do nothing.
```
Comment (AaronRobinsonMSFT) [inline src/vm/class.cpp]: I was assuming you would set the `FALSE` here and remove the comment. Feel free to leave it uninitialized above if concerned about forgetting to set it, but I think we have the "use of uninitialized variable" warning on.

File: src/vm/fieldmarshaler.h [hunk 5]
```diff
@@ -230,7 +259,7 @@ VOID FmtValueTypeUpdateCLR(LPVOID pProtectedManagedData, MethodTable *pMT, BYTE
     }
 
 #define ELEMENT_SIZE_IMPL(NativeSize, AlignmentReq)     \
-    UINT32 NativeSizeImpl() const                       \
+    UINT32 NativeSizeCoreImpl() const                       \
```
Comment (jkotas) [inline src/vm/fieldmarshaler.h]: What does the "Core" in the name stand for? It looks odd to have `NativeSizeCoreImpl` when there is no `NativeSizeImpl` anymore.

Also, some comments still refer to `NativeSizeImpl`.

Comment (jkoritzinsky) [reply src/vm/fieldmarshaler.h]: I've refactored the FieldMarshaler "pseudo-vtable" generation to support generating `NativeSize` without having to use `NativeSizeCore/NativeSizeCoreImpl`.

**Pull request merged**